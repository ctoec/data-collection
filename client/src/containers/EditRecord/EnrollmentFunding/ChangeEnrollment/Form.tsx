import React, { useContext, useState } from 'react';
import {
  Form,
  ExpandCard,
  Button,
  FormSubmitButton,
} from '@ctoec/component-library';
import { ChangeEnrollment } from '../../../../shared/payloads';
import { Enrollment, Child, FundingSource } from '../../../../shared/models';
import {
  SiteField,
  EnrollmentStartDateField,
  CareModelField,
  AgeGroupField,
  NewFundingField,
} from '../../../../components/Forms/Enrollment/Fields';
import { ReportingPeriodField } from '../../../../components/Forms/Enrollment/Funding/Fields';
import UserContext from '../../../../contexts/UserContext/UserContext';
import { apiPost } from '../../../../utils/api';
import AuthenticationContext from '../../../../contexts/AuthenticationContext/AuthenticationContext';

type ChangeEnrollmentFormProps = {
  afterSaveSuccess: () => void;
  afterSaveFailure: (err: any) => void;
  child: Child;
  currentEnrollment?: Enrollment;
};

export const ChangeEnrollmentForm: React.FC<ChangeEnrollmentFormProps> = ({
  afterSaveSuccess,
  afterSaveFailure,
  child,
  currentEnrollment,
}) => {
  const { user } = useContext(UserContext);
  const sites = (user?.sites || []).filter(
    (site) => site.organizationId === child.organization?.id
  );

  const { accessToken } = useContext(AuthenticationContext);

  const [loading, setLoading] = useState(false);
  const onSubmit = (updatedData: ChangeEnrollment) => {
    setLoading(true);
    apiPost(`children/${child.id}/change-enrollment`, updatedData, {
      accessToken,
      jsonParse: false,
    })
      .then(afterSaveSuccess)
      .catch(afterSaveFailure)
      .finally(() => setLoading(false));
  };

  const activeFunding = currentEnrollment?.fundings?.find(
    (f) => !f.lastReportingPeriod
  );

  if (activeFunding && !activeFunding?.fundingSpace) {
    return (
      <div>
        <p>
          Add missing enrollment and/or funding information before you can
          change {child.firstName}'s enrollment
        </p>
        <ExpandCard>
          <Button text="Cancel" appearance="outline" />
        </ExpandCard>
      </div>
    );
  }

  return (
    <Form<ChangeEnrollment>
      id="change-enrollment-form"
      className="usa-form"
      data={{ newEnrollment: {} as Enrollment }}
      onSubmit={onSubmit}
    >
      <SiteField<ChangeEnrollment>
        sites={sites}
        enrollmentAccessor={(data) => data.at('newEnrollment')}
      />
      <EnrollmentStartDateField<ChangeEnrollment>
        enrollmentAccessor={(data) => data.at('newEnrollment')}
      />
      <CareModelField<ChangeEnrollment>
        enrollmentAccessor={(data) => data.at('newEnrollment')}
      />
      <AgeGroupField<ChangeEnrollment>
        enrollmentAccessor={(data) => data.at('newEnrollment')}
      />
      <NewFundingField<ChangeEnrollment>
        fundingAccessor={(data) =>
          data.at('newEnrollment').at('fundings').at(0)
        }
        getEnrollment={(data) => data.at('newEnrollment').value}
        orgId={child.organization.id}
      />

      {!!currentEnrollment && activeFunding && (
        <>
          <h3>Previous enrollment</h3>
          <ReportingPeriodField<ChangeEnrollment>
            accessor={(data) =>
              data.at('oldEnrollment').at('funding').at('lastReportingPeriod')
            }
            fundingSource={activeFunding.fundingSpace?.source as FundingSource} // This is known to have a value (check on line 58)
            isLast={true}
            optional={true}
            label={`Last reporting period for current ${activeFunding.fundingSpace?.source} - ${activeFunding.fundingSpace?.time} funding`}
          />
        </>
      )}

      <ExpandCard>
        <Button text="Cancel" appearance="outline" />
      </ExpandCard>
      <FormSubmitButton
        text={loading ? 'Changing enrollment...' : 'Change enrollment'}
        disabled={loading}
      />
    </Form>
  );
};
