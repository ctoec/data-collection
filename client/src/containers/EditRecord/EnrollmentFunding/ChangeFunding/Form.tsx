import React, { useContext, useState } from 'react';
import { Form, Button, FormSubmitButton } from '@ctoec/component-library';
import { ChangeFunding } from '../../../../shared/payloads';
import { NewFundingField } from '../../../../components/Forms/Enrollment/Fields';
import { ReportingPeriodField } from '../../../../components/Forms/Enrollment/Funding/Fields';
import { Enrollment } from '../../../../shared/models';
import { apiPost } from '../../../../utils/api';
import AuthenticationContext from '../../../../contexts/AuthenticationContext/AuthenticationContext';

type ChangeFundingFormProps = {
  orgId: number;
  afterSaveSuccess: () => void;
  afterSaveFailure: (err: any) => void;
  changeType: 'start' | 'end';
  enrollment: Enrollment;
  hideForm: () => void;
};

export const ChangeFundingForm: React.FC<ChangeFundingFormProps> = ({
  orgId,
  afterSaveSuccess,
  afterSaveFailure,
  changeType,
  enrollment,
  hideForm,
}) => {
  const [loading, setLoading] = useState(false);
  const { accessToken } = useContext(AuthenticationContext);
  const onSubmit = (updatedData: ChangeFunding) => {
    setLoading(true);
    apiPost(`enrollments/${enrollment.id}/change-funding`, updatedData, {
      accessToken,
      jsonParse: false,
    })
      .then(afterSaveSuccess)
      .catch(afterSaveFailure)
      .finally(() => setLoading(false));
  };

  const activeFunding = enrollment.fundings?.find(
    (f) => !f.lastReportingPeriod
  );

  return (
    <Form<ChangeFunding>
      id="change-funding-form"
      className="usa-form"
      data={{}}
      onSubmit={onSubmit}
    >
      {changeType === 'start' && (
        <NewFundingField<ChangeFunding>
          fundingAccessor={(data) => data.at('newFunding')}
          getEnrollment={() => enrollment}
          orgId={orgId}
        />
      )}
      {!!activeFunding && (
        <ReportingPeriodField<ChangeFunding>
          accessor={(data) => data.at('oldFunding').at('lastReportingPeriod')}
          fundingSource={activeFunding.fundingSpace.source}
          isLast={true}
          label={`Last reporting period for current ${activeFunding.fundingSpace?.source} - ${activeFunding.fundingSpace?.time} funding`}
          optional={changeType === 'start'}
        />
      )}
      <Button text="Cancel" appearance="outline" onClick={hideForm} />
      <FormSubmitButton
        text={
          loading
            ? changeType === 'end'
              ? 'Ending funding...'
              : 'Changing funding...'
            : changeType === 'end'
            ? 'End current funding'
            : 'Change current funding'
        }
        disabled={loading}
      />
    </Form>
  );
};
