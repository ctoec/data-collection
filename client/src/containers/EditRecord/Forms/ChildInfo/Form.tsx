import React, { useContext, useState } from 'react';
import {
  SasidField,
  FirstNameField,
  MiddleNameField,
  LastNameField,
  SuffixField,
  DateOfBirthField,
  BirthCertificateFormFieldSet,
  RaceField,
  EthnicityField,
  GenderField,
  FosterCheckbox,
} from './Fields';
import { Form, FormSubmitButton } from '@ctoec/component-library';
import { EditFormProps } from '../types';
import AuthenticationContext from '../../../../contexts/AuthenticationContext/AuthenticationContext';
import { Child } from '../../../../shared/models';
import { apiPut } from '../../../../utils/api';
import useIsMounted from '../../../../hooks/useIsMounted';

// import { useFocusFirstError } from '../../../hooks/useFocusFirstError';

export const ChildInfoForm = ({ child, onSuccess }: EditFormProps) => {
  const { accessToken } = useContext(AuthenticationContext);
  const isMounted = useIsMounted();
  const [saving, setSaving] = useState(false);
  // TODO: HOW ARE WE HANDLING ERRORS?
  // const [error, setError] = useState(null);

  // Focus should automatically be on first error on page
  // useFocusFirstError([error]);

  if (!child) {
    throw new Error('Child info rendered without child');
  }

  const onFormSubmit = (_child: Child) => {
    setSaving(true);
    apiPut(`children/${child.id}`, _child, { accessToken })
      .then(() => onSuccess())
      .catch((err) => {
        console.log(err);
      })
      .finally(() => (isMounted() ? setSaving(false) : null));
  };

  return (
    <Form<Child>
      className="ChildInfoForm usa-form"
      data={child}
      onSubmit={onFormSubmit}
      noValidate
      autoComplete="off"
    >
      <div className="grid-row grid-gap">
        <div className="mobile-lg:grid-col-12">
          <SasidField />
        </div>
        <div className="mobile-lg:grid-col-9">
          <FirstNameField />
        </div>
        <div className="mobile-lg:grid-col-9">
          <MiddleNameField />
        </div>
        <div className="display-flex flex-row flex-align-end grid-row grid-gap">
          <div className="mobile-lg:grid-col-9">
            <LastNameField />
          </div>
          <div className="mobile-lg:grid-col-3">
            <SuffixField />
          </div>
        </div>
      </div>
      <h2>Date of birth</h2>
      <DateOfBirthField />
      <h2>Birth certificate</h2>
      <BirthCertificateFormFieldSet />
      <h2>Race</h2>
      <RaceField />
      <h2>Ethnicity</h2>
      <EthnicityField />
      <h2>Gender</h2>
      <GenderField />
      <h2>Foster</h2>
      <FosterCheckbox />
      <FormSubmitButton
        text={saving ? 'Saving...' : 'Save'}
        disabled={saving}
      />
    </Form>
  );
};
