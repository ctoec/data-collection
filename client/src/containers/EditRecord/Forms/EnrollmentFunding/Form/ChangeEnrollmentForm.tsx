import React, { useContext, useState, useEffect } from 'react';
import AuthenticationContext from '../../../../../contexts/AuthenticationContext/AuthenticationContext';
import {
  Card,
  ExpandCard,
  Button,
  CardExpansion,
  Form,
  FormSubmitButton,
  Alert,
} from '@ctoec/component-library';
import { ChangeEnrollment } from '../../../../../shared/payloads';
import {
  Enrollment,
  Site,
  ReportingPeriod,
  FundingSpace,
} from '../../../../../shared/models';
import { apiPost } from '../../../../../utils/api';
import {
  SiteField,
  AgeGroupField,
  EnrollmentStartDateField,
  EnrollmentEndDateField,
  ReportingPeriodField,
  FundingField,
} from '../Fields';

type ChangeEnrollmentFormProps = {
  childName: string;
  childId: string;
  currentEnrollment?: Enrollment;
  reportingPeriods: ReportingPeriod[];
  fundingSpaces: FundingSpace[];
  sites: Site[];
  refetchChild: () => void;
};

/**
 * Component for gathering user input to change child's Enrollment.
 * Uses a ChangeEnrollment data object to enable the user to provide
 * enrollment end date and funding last reporting period for previously current data.
 */
export const ChangeEnrollmentForm: React.FC<ChangeEnrollmentFormProps> = ({
  childName,
  childId,
  currentEnrollment,
  reportingPeriods,
  fundingSpaces,
  sites,
  refetchChild,
}) => {
  const { accessToken } = useContext(AuthenticationContext);
  const [closeCard, setCloseCard] = useState(false);
  const [error, setError] = useState<string>();
  const [loading, setLoading] = useState(false);

  // Explicitly don't want `closeCard` as a dep, as this
  // needs to be triggered on render caused by child refetch
  // (not only when closeCard changes)
  // eslint-disable-next-line react-hooks/exhaustive-deps
  useEffect(() => {
    if (closeCard) setCloseCard(false);
  });

  const onSubmit = (changeEnrollment: ChangeEnrollment) => {
    setLoading(true);
    apiPost(`children/${childId}/change-enrollment`, changeEnrollment, {
      accessToken,
      jsonParse: false,
    })
      .then(() => {
        setError(undefined);
        setCloseCard(true);
        refetchChild();
      })
      .catch((err) => {
        console.log(err);
        setError(
          'Unable to change enrollment. Make sure all required information is provided'
        );
      })
      .finally(() => setLoading(false));
  };

  const activeFunding = (currentEnrollment?.fundings || []).find(
    (f) => !f.lastReportingPeriod
  );
  return (
    <Card forceClose={closeCard}>
      <div className="display-flex flex-justify">
        <h2 className="header-normal font-heading-lg">
          Has {childName}'s age group and/or site changed?
        </h2>
        <ExpandCard>
          <Button text="Change enrollment" appearance="outline" />
        </ExpandCard>
      </div>
      <CardExpansion>
        <h3 className="margin-top-2 margin-bottom-2 font-heading-lg">
          New enrollment
        </h3>
        {error && <Alert type="error" text={error} />}
        <Form<ChangeEnrollment>
          id="change-enrollment-form"
          className="usa-form"
          data={{ newEnrollment: {} as Enrollment }}
          onSubmit={onSubmit}
        >
          <h4>New enrollment</h4>
          <h4 className="font-heading-md margin-bottom-0">Site</h4>
          <SiteField<ChangeEnrollment>
            sites={sites}
            accessor={(data) => data.at('newEnrollment').at('site')}
          />
          <h4 className="font-heading-md margin-bottom-0">Start date</h4>
          <EnrollmentStartDateField<ChangeEnrollment>
            accessor={(data) => data.at('newEnrollment').at('entry')}
          />
          <h4 className="font-heading-md margin-bottom-0">Age group</h4>
          <AgeGroupField<ChangeEnrollment>
            accessor={(data) => data.at('newEnrollment').at('ageGroup')}
          />

          <FundingField<ChangeEnrollment>
            fundingAccessor={(data) =>
              data.at('newEnrollment').at('fundings').at(0)
            }
            getEnrollment={(data) => data.at('newEnrollment').value}
            fundingSpaces={fundingSpaces}
            reportingPeriods={reportingPeriods}
          />

          {!!currentEnrollment && (
            <>
              <h4>Previous enrollment</h4>
              <h4 className="font-heading-md margin-bottom-0">End date</h4>
              <EnrollmentEndDateField<ChangeEnrollment>
                accessor={(data) => data.at('oldEnrollment').at('exitDate')}
                optional={true}
              />
              {activeFunding && (
                <ReportingPeriodField<ChangeEnrollment>
                  accessor={(data) =>
                    data
                      .at('oldEnrollment')
                      .at('funding')
                      .at('lastReportingPeriod')
                  }
                  reportingPeriods={reportingPeriods.filter(
                    (rp) => rp.type === activeFunding.fundingSpace?.source
                  )}
                  isLast={true}
                  optional={true}
                />
              )}
            </>
          )}

          <ExpandCard>
            <Button text="Cancel" appearance="outline" />
          </ExpandCard>
          <FormSubmitButton
            text={loading ? 'Chaging enrollment...' : 'Change enrollment'}
            disabled={loading}
          />
        </Form>
      </CardExpansion>
    </Card>
  );
};
