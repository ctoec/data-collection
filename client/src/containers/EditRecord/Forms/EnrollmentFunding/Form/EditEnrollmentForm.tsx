import React, { useContext, useEffect, useState } from 'react';
import {
  Card,
  InlineIcon,
  ExpandCard,
  Button,
  TextWithIcon,
  Pencil,
  CardExpansion,
  Form,
  FormSubmitButton,
  Alert,
} from '@ctoec/component-library';
import { Enrollment } from '../../../../../shared/models';
import { EnrollmentEndDateField, EnrollmentStartDateField } from '../Fields';
import { apiPut } from '../../../../../utils/api';
import AuthenticationContext from '../../../../../contexts/AuthenticationContext/AuthenticationContext';

type EditEnrollmentFormProps = {
  enrollment: Enrollment;
  isCurrent?: boolean;
  refetchChild: () => void;
};

/**
 * Component for displaying and editing an existing Enrollment.
 * Does not enable ageGroup to be updated, as this invalidates
 * all funding information, and should instead be handled by deleting
 * and creating a new Enrollment object.
 */
export const EditEnrollmentForm: React.FC<EditEnrollmentFormProps> = ({
  enrollment,
  isCurrent = false,
  refetchChild,
}) => {
  const { accessToken } = useContext(AuthenticationContext);
  const [closeCard, setCloseCard] = useState(false);
  const [error, setError] = useState<string>();
  const [loading, setLoading] = useState(false);

  // Explicitly don't want `closeCard` as a dep, as this
  // needs to be triggered on render caused by child refetch
  // to make form re-openable
  // (not only when closeCard changes)
  // eslint-disable-next-line react-hooks/exhaustive-deps
  useEffect(() => {
    if (closeCard) setCloseCard(false);
  });

  const onSubmit = (updatedEnrollment: Enrollment) => {
    setLoading(true);
    apiPut(`enrollments/${enrollment.id}`, updatedEnrollment, {
      accessToken,
    })
      .then(() => {
        setError(undefined);
        setCloseCard(true);
        refetchChild();
      })
      .catch((err) => {
        setError('Unable to edit enrollment');
      })
      .finally(() => setLoading(false));
  };

  return (
    <Card
      key={enrollment.id}
      appearance={isCurrent ? 'primary' : 'secondary'}
      forceClose={closeCard}
    >
      <div className="display-flex flex-justify">
        <div className="flex-1">
          <p>Site</p>
          <p className="text-bold">{enrollment.site.name}</p>
        </div>
        <div className="flex-1">
          <p>Age group</p>
          <p className="text-bold">
            {enrollment.ageGroup || InlineIcon({ icon: 'incomplete' })}
          </p>
        </div>
        <div className="flex-2">
          <p>Enrollment dates</p>
          <p className="text-bold">
            {enrollment.entry
              ? enrollment.entry.format('MM/DD/YYYY')
              : InlineIcon({ icon: 'incomplete' })}{' '}
            -{' '}
            {enrollment.exit ? enrollment.exit.format('MM/DD/YYYY') : 'present'}
          </p>
        </div>
        <ExpandCard>
          <Button
            text={<TextWithIcon text="Edit" Icon={Pencil} />}
            appearance="unstyled"
          />
        </ExpandCard>
      </div>
      <CardExpansion>
        {error && <Alert type="error" text={error} />}
        <Form<Enrollment>
          id={`edit-enrollment-${enrollment.id}`}
          className="usa-form"
          data={enrollment}
          onSubmit={onSubmit}
        >
          <h3 className="font-heading-md margin-bottom-0">Start date</h3>
          <EnrollmentStartDateField<Enrollment>
            accessor={(data) => data.at('entry')}
          />
          {enrollment.exit && (
            <>
              <h3 className="font-heading-md margin-bottom-0">End date</h3>
              <EnrollmentEndDateField<Enrollment>
                accessor={(data) => data.at('exit')}
              />
            </>
          )}

          <ExpandCard>
            <Button text="Cancel" appearance="outline" />
          </ExpandCard>
          <FormSubmitButton
            text={loading ? 'Saving...' : 'Save'}
            disabled={loading}
          />
        </Form>
      </CardExpansion>
    </Card>
  );
};
