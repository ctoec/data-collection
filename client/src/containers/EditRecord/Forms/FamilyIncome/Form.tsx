import React, { useState } from 'react';
import { Button, Card } from '@ctoec/component-library';
import { propertyDateSorter } from '../../../../utils/dateSorter';
import { IncomeFormProps } from './Fields/Common';
import { EditDeterminationForm } from './EditDeterminationForm';
import { RedeterminationForm } from './RedeterminationForm';

/**
 * The main form rendered in the EditRecord TabNav that allows a user
 * to update the income determination for a given Child record
 * object. Updates are performed on individual income determinations
 * before being re-persisted to the database, but these updates are
 * passed off to the accessible forms (Edit and Redetermine) reached
 * through this page.
 */
export const FamilyIncomeForm: React.FC<IncomeFormProps> = ({
  familyId,
  determinations,
  refetchChild,
}) => {
  const [showNew, setShowNew] = useState(false);
  const [isNew, setIsNew] = useState(false);

  // Tracking variables for easy reference in element creation
  const sortedDeterminations = (determinations || []).sort((a, b) =>
    propertyDateSorter(a, b, (det) => det.determinationDate, true)
  );
  const currentDetermination = sortedDeterminations[0];
  const pastDeterminations = sortedDeterminations.slice(1);

  return (
    <>
      {showNew && (
        <>
          <h2 className="font-sans-md margin-top-2 margin-bottom-2">
            New income determination
          </h2>
          <Card>
            <RedeterminationForm
              familyId={familyId}
              refetchChild={refetchChild}
            />
          </Card>
        </>
      )}
      <div className="margin-top-1">
        <div className="display-flex align-center">
          <h2 className="font-sans-md margin-top-2 margin-bottom-2">
            {currentDetermination
              ? 'Current income determination'
              : 'No income information on record'}
          </h2>
          &nbsp;&nbsp;&nbsp;
          {!showNew && (
            <Button
              text="Redetermine income"
              appearance="unstyled"
              onClick={() => {
                setShowNew(true);
                setIsNew(true);
              }}
            />
          )}
        </div>
        <div>
          {currentDetermination && (
            <EditDeterminationForm
              determination={currentDetermination}
              familyId={familyId}
              isCurrent={true}
              isNew={isNew}
              refetchChild={refetchChild}
            />
          )}
        </div>

        {pastDeterminations.length > 0 && (
          <>
            <div className="display-flex align-center">
              <h2 className="font-sans-md margin-top-2 margin-bottom-2">
                Past income determinations
              </h2>
            </div>
            <div>
              {pastDeterminations.map((determination) => (
                <EditDeterminationForm
                  key={determination.id}
                  determination={determination}
                  familyId={familyId}
                  isCurrent={false}
                  isNew={false}
                  refetchChild={refetchChild}
                />
              ))}
            </div>
          </>
        )}
      </div>
    </>
  );
};
