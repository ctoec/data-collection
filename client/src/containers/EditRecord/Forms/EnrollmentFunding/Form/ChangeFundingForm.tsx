import React, { useContext, useState, useEffect } from 'react';
import {
  Enrollment,
  FundingSpace,
  ReportingPeriod,
  Funding,
} from '../../../../../shared/models';
import AuthenticationContext from '../../../../../contexts/AuthenticationContext/AuthenticationContext';
import { ChangeFunding } from '../../../../../shared/payloads/ChangeFunding';
import { apiPost } from '../../../../../utils/api';
import {
  Card,
  Button,
  Form,
  FormSubmitButton,
  Alert,
} from '@ctoec/component-library';
import { ReportingPeriodField } from '../Fields/Funding/ReportingPeriod';
import { FundingField } from '../Fields/Funding/NewFunding';

type ChangeFundingFormProps = {
  fundingSpaces: FundingSpace[];
  reportingPeriods: ReportingPeriod[];
  enrollment: Enrollment;
  refetchChild: () => void;
};

export const ChangeFundingForm: React.FC<ChangeFundingFormProps> = ({
  fundingSpaces,
  reportingPeriods,
  enrollment,
  refetchChild,
}) => {
  const { accessToken } = useContext(AuthenticationContext);
  const [error, setError] = useState<string>();
  const [loading, setLoading] = useState(false);

  const onSubmit = (changeFunding: ChangeFunding) => {
    setLoading(true);
    apiPost(`enrollments/${enrollment.id}/change-funding`, changeFunding, {
      accessToken,
      jsonParse: false,
    })
      .then(() => {
        setError(undefined);
        setVisibleForm(undefined);
        refetchChild();
      })
      .catch((err) => {
        console.log(err);
        setError(
          'Unable to change funding. Make sure all required information is provided'
        );
      })
      .finally(() => {
        setLoading(false);
      });
  };

  const [visibleForm, setVisibleForm] = useState<'end' | 'start'>();

  const activeFunding = (enrollment.fundings || []).find(
    (f) => !f.lastReportingPeriod
  );
  return (
    <>
      {!!visibleForm && (
        <Card>
          <>
            <h3>
              {visibleForm === 'end'
                ? 'End current funding'
                : 'Start new funding'}
            </h3>
            {error && <Alert type="error" text={error} />}
            <Form<ChangeFunding>
              id="change-funding-form"
              className="usa-form"
              data={{}}
              onSubmit={onSubmit}
            >
              {visibleForm === 'start' && (
                <FundingField
                  reportingPeriods={reportingPeriods}
                  fundingSpaces={fundingSpaces}
                  ageGroup={enrollment.ageGroup}
                  site={enrollment.site}
                  isChangeFunding={true}
                />
              )}
              {!!activeFunding && (
                <ReportingPeriodField
                  reportingPeriods={reportingPeriods.filter(
                    (rp) => rp.type === activeFunding.fundingSpace.source
                  )}
                  isLast={true}
                  isChangeFunding={true}
                />
              )}
              <Button
                text="Cancel"
                appearance="outline"
                onClick={() => setVisibleForm(undefined)}
              />
              <FormSubmitButton
                text={loading ? 'Changing funding...' : 'Change funding'}
                disabled={loading}
              />
            </Form>
          </>
        </Card>
      )}
      <Card>
        <div className="display-flex flex-column flex-align-end">
          <div>
            {activeFunding && (
              <Button
                className="margin-right-4"
                text="End current funding"
                appearance="unstyled"
                onClick={() => setVisibleForm('end')}
              />
            )}
            <Button
              text="Start new funding"
              appearance="unstyled"
              onClick={() => setVisibleForm('start')}
            />
          </div>
        </div>
      </Card>
    </>
  );
};
