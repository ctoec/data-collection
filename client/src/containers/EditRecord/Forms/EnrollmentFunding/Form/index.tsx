import React, { useContext, useEffect, useState } from 'react';
import {
  Site,
  Enrollment,
  FundingSpace,
  ReportingPeriod,
} from '../../../../../shared/models';
import { apiGet } from '../../../../../utils/api';
import AuthenticationContext from '../../../../../contexts/AuthenticationContext/AuthenticationContext';
import { ChangeEnrollmentForm } from './ChangeEnrollmentForm';
import { EditEnrollmentForm } from './EditEnrollmentForm';
import { EditFundingForm } from './EditFundingForm';
import { ChangeFundingForm } from './ChangeFundingForm';

type EnrollmentFundingFormProps = {
  reportingPeriods: ReportingPeriod[];
  enrollments: Enrollment[];
  childName: string;
  childId: string;
  refetchChild: () => void;
};
export const EnrollmentFundingForm: React.FC<EnrollmentFundingFormProps> = ({
  reportingPeriods,
  enrollments,
  childName,
  childId,
  refetchChild,
}) => {
  const { accessToken } = useContext(AuthenticationContext);

  // Get site options for new enrollments
  const [sites, setSites] = useState<Site[]>([]);
  useEffect(() => {
    apiGet('sites', { accessToken }).then((_sites) => setSites(_sites));
  }, [accessToken]);

  // Get fundingSpaces for new fundings
  const [fundingSpaces, setFundingSpaces] = useState<FundingSpace[]>([]);
  useEffect(() => {
    apiGet('funding-spaces', { accessToken }).then((_fundingSpaces) =>
      setFundingSpaces(_fundingSpaces)
    );
  }, [accessToken]);

  // Separate enrollments into current (no end date) and past
  // (with end date). Either may not exist
  const currentEnrollment = enrollments.find((e) => !e.exit);
  const pastEnrollments = currentEnrollment
    ? enrollments.filter((e) => e.id !== currentEnrollment.id)
    : enrollments;

  return (
    <>
      <ChangeEnrollmentForm
        reportingPeriods={reportingPeriods}
        fundingSpaces={fundingSpaces}
        childName={childName}
        currentEnrollment={currentEnrollment}
        childId={childId}
        sites={sites}
        refetchChild={refetchChild}
      />
      {currentEnrollment && (
        <>
          <h2 className="margin-top-4 margin-bottom-2 font-heading-md">
            Current enrollment
          </h2>
          <EditEnrollmentForm
            isCurrent={true}
            enrollment={currentEnrollment}
            refetchChild={refetchChild}
          />
          {currentEnrollment.fundings?.map((funding) => (
            <EditFundingForm
              key={funding.id}
              isCurrent={true}
              reportingPeriods={reportingPeriods}
              fundingSpaces={fundingSpaces}
              funding={funding}
              enrollment={currentEnrollment}
              refetchChild={refetchChild}
            />
          ))}
          <ChangeFundingForm
            fundingSpaces={fundingSpaces}
            reportingPeriods={reportingPeriods}
            enrollment={currentEnrollment}
            refetchChild={refetchChild}
          />
        </>
      )}
      {!!pastEnrollments.length && (
        <>
          <h2 className="margin-top-4 margin-bottom-2 font-heading-md">
            Past enrollments
          </h2>
          {pastEnrollments.map((enrollment) => (
            <>
              <EditEnrollmentForm
                key={enrollment.id}
                enrollment={enrollment}
                refetchChild={refetchChild}
              />
              {enrollment.fundings?.map((funding) => (
                <EditFundingForm
                  key={funding.id}
                  reportingPeriods={reportingPeriods}
                  fundingSpaces={fundingSpaces}
                  funding={funding}
                  enrollment={enrollment}
                  refetchChild={refetchChild}
                />
              ))}
            </>
          ))}
        </>
      )}
    </>
  );
};
