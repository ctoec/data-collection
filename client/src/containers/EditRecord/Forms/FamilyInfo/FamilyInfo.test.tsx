import React from 'react';
import {
  snapshotTestHelper,
  accessibilityTestHelper,
} from '../../../../testHelpers';
import { FamilyInfoForm } from './Form';
import { Family } from '../../../../shared/models';

describe('EditRecord', () => {
  describe('FamilyInfo', () => {
    snapshotTestHelper(
      <FamilyInfoForm family={{} as Family} refetchChild={jest.fn()} />
    );
    accessibilityTestHelper(
      <FamilyInfoForm family={{} as Family} refetchChild={jest.fn()} />
    );
  });
});
