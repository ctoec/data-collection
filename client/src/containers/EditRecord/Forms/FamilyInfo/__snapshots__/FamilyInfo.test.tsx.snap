// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditRecord FamilyInfo matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="grid-container margin-top-2"
  >
    <h1
      class="grid-row"
    >
      Family Address
    </h1>
    <div
      class="grid-row margin-top-2 text-light"
    >
      Information is required unless otherwise specified.
    </div>
    <form
      autocomplete="off"
      class="FamilyInfoForm"
      novalidate=""
    >
      <h2
        class="grid-row margin-top-4"
      >
        Address
      </h2>
      <fieldset
        class="grid-gap grid-row usa-fieldset display-inline-block"
        id="family-address"
        role="presentation"
      >
        <legend
          id="fieldset-legend-family-address"
        >
          <span
            class="usa-sr-only"
          >
            Address
          </span>
        </legend>
        <div
          class="grid-gap grid-col"
        >
          <div
            class="grid-col-12 margin-top-2"
          >
            <div
              class=" usa-form-group"
            >
              <label
                class="usa-label"
                for="street-address"
              >
                Street address 
              </label>
              <input
                aria-required="true"
                class="usa-input"
                id="street-address"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class="grid-row grid-gap"
          >
            <div
              class="grid-col-8 display-inline-block"
            >
              <div
                class=" usa-form-group"
              >
                <label
                  class="usa-label"
                  for="town"
                >
                  Town 
                </label>
                <input
                  aria-required="true"
                  class="usa-input"
                  id="town"
                  type="text"
                  value=""
                />
              </div>
            </div>
            <div
              class="grid-col-4 display-inline-block"
            >
              <div
                class="usa-form-group"
              >
                <div
                  class=""
                />
                <label
                  class="usa-label"
                  for="state"
                >
                  State
                </label>
                <select
                  aria-required="true"
                  class="usa-select"
                  id="state"
                  name="state"
                >
                  <option
                    value=""
                  >
                    - Select -
                  </option>
                  <option
                    value="CT"
                  >
                    CT
                  </option>
                  <option
                    value="MA"
                  >
                    MA
                  </option>
                  <option
                    value="NY"
                  >
                    NY
                  </option>
                  <option
                    value="RI"
                  >
                    RI
                  </option>
                </select>
              </div>
            </div>
          </div>
          <div
            class="grid-col-6"
          >
            <div
              class=" usa-form-group"
            >
              <label
                class="usa-label"
                for="zip"
              >
                Zip code 
              </label>
              <input
                aria-required="true"
                class="usa-input"
                id="zip"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
      </fieldset>
      â€‹
      <div
        class="grid-row margin-top-4"
      >
        <div
          class="usa-checkbox"
        >
          <input
            class="usa-checkbox__input"
            id="homelessness"
            type="checkbox"
            value="homelessness"
          />
          <label
            class="usa-checkbox__label"
            for="homelessness"
          >
            Family has experienced homelessness / housing insecurity within the last year
          </label>
        </div>
      </div>
      <div
        class="grid-row margin-top-2"
      >
        <input
          class="usa-button"
          type="submit"
          value="Save edits"
        />
      </div>
    </form>
  </div>
</DocumentFragment>
`;
